### PR TITLE
 #187300249  Seller CRUD operation over their  products

### PR DESCRIPTION
**Pull request by emmanueltct**
#### What does this PR do?
- This feature implements the seller crud operation over their products
#### Description of Task to be completed?
- to add product item user should be authenticated and has role of seller
- Seller could be able to read, update , and delete his/her products
- no seller could have two products of same name 
- each product must have at least between 4 and 8 pictures
#### How should this be manually tested?
- Should be tested on swagger
- use a postman to test it locally
#### What are the relevant pivotal tracker/Trello stories?
#187300249

### CODE REVIEW

**Review by ErnestTchami — APPROVED** [review]

**Review by kanu-cast — APPROVED** [review]

File: src/config/cloudinary.config.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { v2 as cloudinary } from "cloudinary";
+import dotenv from "dotenv";
+
+dotenv.config();
+
+cloudinary.config({
+  cloud_name: "dbbv6uvfu",
+  api_key: "496227742425544",
+  api_secret: "-3JAdSg3tFY2li0_fyRcy3ybVU0"
```
Comment (Eli250) [inline src/config/cloudinary.config.ts]: You should always keep secrets in environment variables. Do not expose them on the client.

File: src/controllers/product.controllers.ts [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        productPictures: urls,
+        expireDate
+      });
+      res
+        .status(201)
+        .json({ message: "Product item is successful created", product });
+    } else {
+      return res
+        .status(403)
+        .json({ error: "Product item should have 4 up to 10 images" });
```
Comment (Eli250) [inline src/controllers/product.controllers.ts]: This should be done via validation middleware

File: src/controllers/product.controllers.ts [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    const userId: string = logedUser.dataValues.id as string;
+    const product: ProductAttributes | any = await Product.findOne({
+      where: {
+        id: req.params.product_id,
+        sellerId: userId
+      }
+    });
+
+    if (product) {
+      if (req.body.productName) {
```
Comment (Eli250) [inline src/controllers/product.controllers.ts]: Remove these ifs here. You need a middleware that checks if the product exists before you start any process.
Second you need a validation middleware that validates all fields.

File: src/controllers/product_category.controllers.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import { Request, Response } from "express";
```
Comment (Eli250) [inline src/controllers/product_category.controllers.ts]: File naming does not follow the conventions set in your codebase.

File: src/middlewares/product_category.middlewares.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import { NextFunction, Request, Response } from "express";
```
Comment (Eli250) [inline src/middlewares/product_category.middlewares.ts]: File naming does not follow the conventions. Change 'product_category' to productCategory.

File: src/middlewares/product_category.middlewares.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+import {
+  productCategorySchema,
+  productCategoryUpdateSchema
+} from "../validations/product_categories.validations";
+import ProductCategory from "../models/product_category";
+
+export const isValidCategoryInsert = (
+  req: Request,
```
Comment (Eli250) [inline src/middlewares/product_category.middlewares.ts]: This function should be under validations directory not in a middlewares.

File: src/middlewares/product_category.middlewares.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    const result = productCategorySchema.parse(req.body);
+    if (result) {
+      next();
+    }
+  } catch (error: any) {
+    return res.status(400).json({ error: error.errors[0].message });
+  }
+};
+
+export const isValidCategoryUpdate = (
```
Comment (Eli250) [inline src/middlewares/product_category.middlewares.ts]: This function should be under validations not middlewares.

File: src/middlewares/product_category.middlewares.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  }
+};
+
+export const isExistProductCategory = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    let cat_id: string;
```
Comment (Eli250) [inline src/middlewares/product_category.middlewares.ts]: This variable naming is wrong.

File: src/middlewares/sellerproduct.middlewares.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
+import { NextFunction, Request, Response } from "express";
```
Comment (Eli250) [inline src/middlewares/sellerproduct.middlewares.ts:1]: These functionalities should be under product validation or middlewares, there is no need to create another file.

File: src/middlewares/sellerproduct.middlewares.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+import multer, { MulterError } from "multer";
+import Product from "../models/Product";
+import {
+  productSchema,
+  productUpdateSchema
+} from "../validations/seller.product.validation";
+import User from "../models/user";
+import ProductCategory from "../models/product_category";
+
+export const isValidItem = (
```
Comment (Eli250) [inline src/middlewares/sellerproduct.middlewares.ts]: This should be under product validation file not middleware.

File: src/middlewares/sellerproduct.middlewares.ts [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    const result = productSchema.parse(req.body);
+    if (result) {
+      next();
+    }
+  } catch (error: any) {
+    return res.status(400).json({ error: error.errors[0].message });
+  }
+};
+
+export const isValidUpdate = (
```
Comment (Eli250) [inline src/middlewares/sellerproduct.middlewares.ts]: This should alse be in validations not middlewares.

File: src/middlewares/sellerproduct.middlewares.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+    const product = await Product.findOne({
+      where: {
+        sellerId: userId,
+        productName: req.body.productName
+      }
+    });
+    if (product) {
+      return res.status(409).json({
+        message: "This item is alread Exist , please update it",
```
Comment (Eli250) [inline src/middlewares/sellerproduct.middlewares.ts]: Change the message to "This item already exists"

File: src/middlewares/sellerproduct.middlewares.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    next();
+  } catch (error) {
+    return res.status(500).json({ error: "Internal server error" });
+  }
+};
+
+export const isValidImage = async (
+  err: MulterError | Error,
+  req: any,
```
Comment (Eli250) [inline src/middlewares/sellerproduct.middlewares.ts:87]: Don't use "any"

File: src/models/product_category.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import { DataTypes, Model } from "sequelize";
```
Comment (Eli250) [inline src/models/product_category.ts:1]: File naming is wrong. Use "productCategory.ts" instead.

File: src/routes/product_category.routes.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
+import express from "express";
```
Comment (Eli250) [inline src/routes/product_category.routes.ts]: Change file name to "productCategory.routes.ts"

**Review by Eli250 — CHANGES_REQUESTED** [review]

**Review by Eli250 — APPROVED** [review]

**Pull request merged**